specs/eventing/control-plane.md

```diff
@@ -0,0 +1,985 @@
+HTTP formats. Before sending an HTTP response, the Channel MUST durably enqueue
+the event (be able to deliver with retry without receiving the event again).
+
+For each event received by the Channel, the Channel MUST deliver the event to
+each associated Subscription **at least once** (where "associated" means a
+Subscription with a `spec.channel` which references the Channel). If the
+Subscription has a `Ready` condition of `true` when the event is evaluated, the
+Channel MUST forward the event as described in
+[event delivery as described below](#event-delivery). The Channel MAY also
+forward events to associated Subscriptions for with the `Ready` condition is not
+currently `true`. (One example: a Subscription which is in the process of being
+programmed in the Channel data plane might receive _some_ events before the data
+plane programming was complete and the Subscription was updated to set the
+`Ready` condition to `true`.)
+
+If multiple Subscriptions with the same destination are associated with the same
+Channel, each Subscription MUST generate one delivery attempt per Subscription.
+The implementation MAY attach additional event attributes or other metadata
+distinguishing between these deliveries. The implementation MUST NOT modify the
+event payload in this process.
+
+### Event Delivery
+
+Once a Trigger or Subscription has decided to deliver an event, it MUST do the
+following:
+
+1. Resolve all URLs and delivery options, using the values in `status` for URL
+   resolution.
+
+1. Attempt delivery to the `status.subscriberUri` URL following the
+   [data plane contract](./data-plane.md).
+
+   1. If the event delivery fails with a retryable error, it SHOULD be retried
+      up to `retry` times, following the `backoffPolicy` and `backoffDelay`
+      parameters if specified.
+
+1. If the delivery attempt is successful (either the original request or a
+   retry) and no event is returned, the event delivery is complete.
+
+1. If the delivery attempt is successful (either the original request or a
+   retry) and an event is returned in the reply, the reply event will be
+   delivered to the `status.replyUri` destination (for Subscriptions) or added
+   to the Broker for processing (for Triggers). If `status.replyUri` is not
+   present in the Subscription, the reply event is be dropped.
+
+   1. If delivery of the reply event fails with a retryable error, the delivery
+      of the reply event to the reply destination SHOULD be retried up to
+      `retry` times, following the `backoffPolicy` and `backoffDelay` parameters
+      if specified.
+
+1. If all retries are exhausted for either the original delivery or the retry,
+   or if a non-retryable error is received, the event should be delivered to the
+   `deadLetterSink` in the delivery options. If no `deadLetterSink` is
+   specified, the event is dropped.
+
+   The implementation MAY set additional attributes on the event or wrap the
```

Comment:
The wrapper bit is interesting. I think I understand why some folks may want to do that, but won't that hurt interop? I think adding extra metadata would be sufficient. What if an impl decided to randomly add a wrap, or not? Not sure how the wrapper helps the user more than extra metadata fields.

Reply:
Why would we need to convert binary events into JSON? That's an impl choice. They could just store the http request as http headers + binary body.

I'm ok with waiting until a future rev of the spec, but it seems to me that some folks might implement some kind of DLQ processing and would like for that to be portable/interoperable - so knowing how to parse the queue would be necessary.

Reply:
Actually, think more about this... I'd like to suggest that we DROP the idea of wrapping the event. If we allow people to wrap the event then statements like `the event MUST be delivered to the deadLetterSink` is inconsistent. By that I mean, not all destinations for events are the same and I don't see why we would want some event flows to follow a different pattern than others. How things are stored in the DLQ is an impl choice though.